client/connection/manager_test.go

```diff
@@ -78,52 +93,61 @@ func (tc *testContext) TestWithUnknownProviderConnectionIsNotMade() {
 
 func (tc *testContext) TestOnConnectErrorStatusIsNotConnectedAndSessionStartIsNotMarked() {
 	fatalVpnError := errors.New("fatal connection error")
-	tc.fakeOpenVpn.onConnectReturnError = fatalVpnError
+	tc.fakeOpenVpn.onStartReturnError = fatalVpnError
 
 	assert.Error(tc.T(), tc.connManager.Connect(myID, activeProviderID))
 	assert.Equal(tc.T(), statusNotConnected(), tc.connManager.Status())
-
+	assert.True(tc.T(), tc.fakeDialog.closed)
 	assert.False(tc.T(), tc.fakeStatsKeeper.sessionStartMarked)
 }
 
 func (tc *testContext) TestWhenManagerMadeConnectionStatusReturnsConnectedStateAndSessionId() {
 	err := tc.connManager.Connect(myID, activeProviderID)
-	tc.fakeOpenVpn.reportState(openvpn.ConnectedState)
 	assert.NoError(tc.T(), err)
 	assert.Equal(tc.T(), statusConnected("vpn-connection-id"), tc.connManager.Status())
 }
 
 func (tc *testContext) TestWhenManagerMadeConnectionSessionStartIsMarked() {
 	err := tc.connManager.Connect(myID, activeProviderID)
-	tc.fakeOpenVpn.reportState(openvpn.ConnectedState)
 	assert.NoError(tc.T(), err)
 
 	assert.True(tc.T(), tc.fakeStatsKeeper.sessionStartMarked)
 }
 
 func (tc *testContext) TestStatusReportsConnectingWhenConnectionIsInProgress() {
-	tc.connManager.Connect(myID, activeProviderID)
-	assert.Equal(tc.T(), statusConnecting(), tc.connManager.Status())
+	tc.fakeOpenVpn.onStartReportStates = []openvpn.State{}
+	errorChannel := make(chan error, 1)
+
+	go func() {
+		errorChannel <- tc.connManager.Connect(myID, activeProviderID)
+	}()
+
+	waitABit()
+	select {
+	case err := <-errorChannel:
+		assert.Fail(tc.T(), "Connect error not expected: ", err)
+	default:
+		assert.Equal(tc.T(), statusConnecting(), tc.connManager.Status())
+	}
```

Comment:
We can use boolean for that - getting boolean value from variable and setting it are atomic operations AFAIK.
```go
func (tc *testContext) TestStatusReportsConnectingWhenConnectionIsInProgress() {
	tc.fakeOpenVpn.onStartReportStates = []openvpn.State{}

	connected := false
	go func() {
		tc.connManager.Connect(myID, activeProviderID)
		connected = true
	}()

	waitABit()
	assert.Equal(tc.T(), statusConnecting(), tc.connManager.Status())
	assert.False(tc.T(), connected)
}
```

Reply:
```go
func (tc *testContext) TestStatusReportsConnectingWhenConnectionIsInProgress() {
	tc.fakeOpenVpn.onStartReportStates = []openvpn.State{}

	go func() {
		tc.connManager.Connect(myID, activeProviderID)
		assert.Fail(tc.T(), "failed")
	}()

	waitABit()
	assert.Equal(tc.T(), statusConnecting(), tc.connManager.Status())
}
```